eclair-core/src/main/scala/fr/acinq/eclair/balance/BalanceActor.scala

```diff
@@ -0,0 +1,139 @@
+package fr.acinq.eclair.balance
+
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import akka.actor.typed.{ActorRef, Behavior}
+import fr.acinq.bitcoin.ByteVector32
+import fr.acinq.eclair.balance.BalanceActor._
+import fr.acinq.eclair.balance.CheckBalance.GlobalBalance
+import fr.acinq.eclair.balance.Monitoring.{Metrics, Tags}
+import fr.acinq.eclair.blockchain.bitcoind.rpc.ExtendedBitcoinClient
+import fr.acinq.eclair.blockchain.bitcoind.rpc.ExtendedBitcoinClient.Utxo
+import fr.acinq.eclair.channel.HasCommitments
+import fr.acinq.eclair.db.PendingCommandsDb
+import grizzled.slf4j.Logger
+import org.json4s.JsonAST.JInt
+
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+object BalanceActor {
+
+  // @formatter:off
+  sealed trait Command
+  private final case object TickBalance extends Command
+  final case class GetGlobalBalance(replyTo: ActorRef[Future[GlobalBalance]], channels: Map[ByteVector32, HasCommitments]) extends Command
```

Comment:
Indeed it is a bit strange, but we have to manage compatibility between typed ask and the existing api code. The alternative is:

```diff
diff --git a/eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala b/eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala
index 44dd7aff4..ae29fec4c 100644
--- a/eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala
@@ -438,7 +438,8 @@ class EclairImpl(appKit: Kit) extends Eclair with Logging {
   override def globalBalance()(implicit timeout: Timeout): Future[GlobalBalance] = {
     for {
       ChannelsListener.GetChannelsResponse(channels) <- appKit.channelsListener.ask(ref => ChannelsListener.GetChannels(ref))(timeout, appKit.system.scheduler.toTyped)
-      globalBalance <- appKit.balanceActor.ask(res => BalanceActor.GetGlobalBalance(res, channels))(timeout, appKit.system.scheduler.toTyped).flatten
+      globalBalance_try <- appKit.balanceActor.ask(res => BalanceActor.GetGlobalBalance(res, channels))(timeout, appKit.system.scheduler.toTyped)
+      globalBalance = globalBalance_try.get
     } yield globalBalance
   }

diff --git a/eclair-core/src/main/scala/fr/acinq/eclair/balance/BalanceActor.scala b/eclair-core/src/main/scala/fr/acinq/eclair/balance/BalanceActor.scala
index f1963fe37..29b553c7d 100644
--- a/eclair-core/src/main/scala/fr/acinq/eclair/balance/BalanceActor.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/balance/BalanceActor.scala
@@ -1,5 +1,6 @@
 package fr.acinq.eclair.balance

+import akka.actor.Status
 import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
 import akka.actor.typed.{ActorRef, Behavior}
 import fr.acinq.bitcoin.ByteVector32
@@ -22,7 +23,7 @@ object BalanceActor {
   // @formatter:off
   sealed trait Command
   private final case object TickBalance extends Command
-  final case class GetGlobalBalance(replyTo: ActorRef[Future[GlobalBalance]], channels: Map[ByteVector32, HasCommitments]) extends Command
+  final case class GetGlobalBalance(replyTo: ActorRef[Try[GlobalBalance]], channels: Map[ByteVector32, HasCommitments]) extends Command
   private final case class WrappedChannels(wrapped: ChannelsListener.GetChannelsResponse) extends Command
   private final case class WrappedGlobalBalance(wrapped: Try[GlobalBalance]) extends Command
   private final case class WrappedUtxoInfo(wrapped: Try[UtxoInfo]) extends Command
@@ -112,7 +113,7 @@ private class BalanceActor(context: ActorContext[Command],
           Behaviors.same
       }
     case GetGlobalBalance(replyTo, channels) =>
-      replyTo ! CheckBalance.computeGlobalBalance(channels, pendingCommandsDb, extendedBitcoinClient)
+      CheckBalance.computeGlobalBalance(channels, pendingCommandsDb, extendedBitcoinClient) onComplete (replyTo ! _)
       Behaviors.same
     case WrappedUtxoInfo(res) =>
       res match {
```

Reply:
I think this is slightly better. Instead of a `for` comprehension (which restricts us a bit here), you could also simply write the `map/flatMap` yourself and explicitly handle the failure case (which gets rid of the `.get`).

Reply:
🤓 

```scala
globalBalance <- Promise[GlobalBalance]().complete(globalBalance_try).future
```

Reply:
![witchcraft](https://user-images.githubusercontent.com/31281497/124278912-254e9500-db47-11eb-870d-97544f6ebd27.jpeg)

Reply:
Applied in 3ffbe81.